src/string_object.cpp

```diff
@@ -2220,42 +2220,97 @@ Value StringObject::rstrip_in_place(Env *env) {
     return this;
 }
 
-Value StringObject::downcase(Env *env) {
-    auto ary = chars(env)->as_array();
+// This implements checking the case-fold options passed into arguments like
+// downcase, upcase, casecmp, etc and sets a bitfield enum.
+CaseFoldType StringObject::check_case_options(Env *env, Value arg1, Value arg2, CaseFoldType flags) {
+    SymbolObject *turk = "turkic"_s;
+    SymbolObject *lith = "lithuanian"_s;
+    // return for zero arg case
+    if (arg1.is_null() && arg2.is_null())
+        return flags;
+    // two arg case only accepts turkic and lithuanian (in either order)
+    if (!arg1.is_null() && !arg2.is_null()) {
+        if ((arg1 == turk && arg2 == lith) || (arg1 == lith && arg2 == turk)) {
+            return flags | FoldTurkicAzeri | FoldLithuanian;
+        } else {
+            // any other pair of arguments is an error
+            env->raise("ArgumentError", "invalid option");
+        }
+    }
+    // acceptable symbols as options: [turkic lithuanian ascii fold]
+    if (arg1 == "ascii"_s) {
+        return flags | Ascii;
+    } else if (arg1 == "fold"_s) {
+        if ((flags & (Upcase | Downcase)) == Downcase) {
+            flags = flags ^ (Fold | Downcase);
+            return flags;
+        } else {
+            env->raise("ArgumentError", "option :fold only allowed for downcasing");
+        }
+    } else if (arg1 == turk) {
+        return flags | FoldTurkicAzeri;
+    } else if (arg1 == lith) {
+        return flags | FoldLithuanian;
+    } else {
+        env->raise("ArgumentError", "invalid option");
+    }
+    return flags;
+}
+
+Value StringObject::downcase(Env *env, Value arg1, Value arg2) {
+    CaseFoldType flags = check_case_options(env, arg1, arg2, Downcase);
     auto str = new StringObject { "", m_encoding };
-    for (auto c_val : *ary) {
-        auto c_str = c_val->as_string();
-        if (c_str->bytesize() > 1) {
-            str->append(c_str);
+    for (StringView c : *this) {
```

Comment:
Fewer lines than the older method, but maybe there was a reason for the old way.  Decoding/modifying and re-encoding codepoints feels more appropriate in the context of supporting multi-byte encodings like UTF-16/32, etc.

It feels like the case-mapping functions may eventually want to live with the encoding object itself?